# Table of Content


1. **[Importing Modules (Pandas)](#pandas)**
<br><br> 
2. **[Pandas DataFrame](#dataframes)**
<br><br>
3. **[Manipulating DataFrame](#dataframes)**
<br><br>
4. **[Reading Data from Different Sources](#reading_data)**



<a id="pandas"> </a>
# 1. Pandas

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.
<br><br>
                       Pandas is designed for working with tabular data.<br><br>

A module/library in python is simply a way to organize the code, and it contains either python classes or just functions. 

### Quick look at functions

In [8]:
# what are functions and arguments
x,y = 5,7
def add(x,y):
    c = x+y
    return c
print(add(5,7))

12


In [10]:
import cal

cal.add(5,7)

12

In [11]:
cal.cross(5,7)

35

In [2]:
#Explain functions and how import works to students
# import inspect
# print(inspect.getsource(add.ad))

**How to install and import pandas?**<br>
1. Install pandas:<br><br>
`!pip install pandas`<br><br>
2. Import pandas:<br><br>
`import pandas as pd`

In [1]:
#Check the list of base packages
!pip list

Package                       Version
----------------------------- --------------------
alabaster                     0.7.12
anaconda-client               1.11.0
anaconda-navigator            2.3.1
anaconda-project              0.11.1
anyio                         3.5.0
appdirs                       1.4.4
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
arrow                         1.2.2
astroid                       2.11.7
astropy                       5.1
atomicwrites                  1.4.0
attrs                         21.4.0
Automat                       20.2.0
autopep8                      1.6.0
Babel                         2.9.1
backcall                      0.2.0
backports.functools-lru-cache 1.6.4
backports.tempfile            1.0
backports.weakref             1.0.post1
bcrypt                        3.2.0
beautifulsoup4                4.11.1
binaryornot                   0.4.4
bitarray                      2.5.1
bkcharts                      0.2
blac

In [ ]:
# install pandas (one time only)
# !pip install pandas

In [1]:
#import pandas 'library/package/modules'
import pandas as pd

 `as` is used as an alias in pandas. So from now on we will use `pd.` instead of `pandas.` 
 
<br>
<span style="color:crimson">Always use libraries if they are freely available. It saves time, and those codes are already tested, debugged and optimized.</span>

<a id="dataframes"> </a>
# 2. Pandas DataFrames

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (numeric, string, boolean, and so on). <br><br>                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">To read data from a csv file</span>

In [2]:
# read the example.csv file in a dataframe
data1 = pd.read_csv('BMI.csv')
data2 = pd.read_csv('missingdata.csv')

In [4]:
# check the type
type(data1)

pandas.core.frame.DataFrame

In [5]:
type(data2)

pandas.core.frame.DataFrame

On checking the data type, we notice it is read as pandas data frame.

## <span style="color:darkgreen;">To print top & bottom rows of the data</span>

In [6]:
#top 3 rows
data1.head(3)

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

In [7]:
#bottom 3 rows
data1.tail(5)

Age  Weight (in kg)  Height (in m)
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

## <span style="color:darkgreen;">To obtain the dimensions of the data</span>

In [9]:
# get the shape of the dataset
data1.shape

#size of the dataset
size = data1.size
print(size)

69


In [26]:
data1.shape

(23, 3)

## <span style="color:darkgreen;">To know the data types of a data frame</span>

In [12]:
# get all the data types
data1.dtypes

Age                 int64
Weight (in kg)      int64
Height (in m)     float64
dtype: object

In [13]:
data2.dtypes

Age               float64
Weight (in kg)    float64
Height (in m)     float64
dtype: object

We see the data type of each variable.

## <span style="color:darkgreen;">Print more information about the data</span>

In [14]:
# All information together
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             23 non-null     int64  
 1   Weight (in kg)  23 non-null     int64  
 2   Height (in m)   23 non-null     float64
dtypes: float64(1), int64(2)
memory usage: 680.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`. 

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.

In [15]:
# describe your data
data1.describe()

Age  Weight (in kg)  Height (in m)
count  23.000000       23.000000      23.000000
mean   44.521739       58.304348       1.528261
std    20.586557       19.401112       0.227309
min    10.000000       21.000000       1.210000
25%    26.000000       44.000000       1.320000
50%    54.000000       65.000000       1.520000
75%    62.000000       76.000000       1.690000
max    75.000000       89.000000       1.850000

# `.loc` and `.iloc` methods

## <span style="color:darkgreen;">Indexing a dataframe using `.loc`</span>

`DataFrame.loc[]` is label-based method, which means that you have to specify rows and columns based on their row and column labels.
https://www.geeksforgeeks.org/indexing-and-selecting-data-with-pandas/
this is also called slice and dicing of dataset.

In [ ]:
#syntax
#data1.loc[row-start:row-end, column-start:column-end]

In [32]:
data1.loc[0]

Age               45.00
Weight (in kg)    60.00
Height (in m)      1.35
Name: 0, dtype: float64

In [20]:
# get first row with all columns
data1.loc[[0], :]

Age  Weight (in kg)  Height (in m)
0   45              60           1.35

In [21]:
data1.iloc[[0]]

Age  Weight (in kg)  Height (in m)
0   45              60           1.35

In [35]:

data1.loc[[0], ["Age"]]

Age
0   45

In [34]:
# get the value of Age in the 1st row
data1.loc[0, 'Age']

45

## <span style="color:darkgreen;">Selecting multiple rows</span>

In [36]:
# select row 4, 6, 10
data1.iloc[[3,5,9]]

Age  Weight (in kg)  Height (in m)
3   26              65           1.21
5   21              43           1.52
9   32              21           1.52

We use two square brackets since we are passing a list of row numbers to be accessed.

## <span style="color:darkgreen;">Selecting a range of rows</span>

In [37]:
# select rows from 11 to 18
data1.iloc[10:17]

Age  Weight (in kg)  Height (in m)
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21

## <span style="color:darkgreen;">Selecting the first column</span>

In [39]:
data1.loc[:, :'Age']

Age
0    45
1    12
2    54
3    26
4    68
5    21
6    10
7    57
8    75
9    32
10   23
11   34
12   55
13   23
14   56
15   67
16   26
17   56
18   67
19   26
20   68
21   56
22   67

In [28]:
# select 1st column with all rows
data1.iloc[:, [0]]

Age
0    45
1    12
2    54
3    26
4    68
5    21
6    10
7    57
8    75
9    32
10   23
11   34
12   55
13   23
14   56
15   67
16   26
17   56
18   67
19   26
20   68
21   56
22   67

To select the last column we use -1, to select the second last column we use -2

## <span style="color:darkgreen;">Select the first two columns</span>

In [31]:
# select the first two columns with all rows
data1.iloc[: , 0:2]

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

## <span style="color:darkgreen;">Indexing a dataframe using `.iloc`</span>

`DataFrame.iloc[]` is integer position-based, so you have to specify rows and columns by their integer position values (0-based integer position).

**Note:** the row names are numbers 

In [40]:
# using iloc select 'Age' & 'Weight (in kg)'
data1.iloc[:, 0:2]

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

<a id="manipulatingDF"> </a>
# 3. Manipulating a Dataframe

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> CAUTION:<br>
                        1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
                        2. New columns cannot be created with the ` data.BMI ` syntax.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">Adding a new column to the dataframe</span>

columns/variables/features mean the same

In [ ]:
data1[['Age', 'Weight (in kg)']]

Feature Engineering

In [4]:
# create a new column BMI which is given by weight / H**2
data1['BMI'] = data1['Weight (in kg)']/data1['Height (in m)']**2
data1.head()

Age  Weight (in kg)  Height (in m)        BMI
0   45              60           1.35  32.921811
1   12              43           1.21  29.369579
2   54              78           1.50  34.666667
3   26              65           1.21  44.395875
4   68              50           1.32  28.696051

In [5]:
# check the shape of the data
data1.shape

(23, 4)

## <span style="color:darkgreen;">Adding a new row to the dataframe</span>

In [6]:
# add a new row to the data with values [56, 76, 1.69, 26.609713] using .loc method
data1.loc[23] = [56, 76, 1.69, 26.609713]
data1.tail(2)

Age  Weight (in kg)  Height (in m)        BMI
22  67.0            78.0           1.85  22.790358
23  56.0            76.0           1.69  26.609713

We see that a new row number 23 has be added to the data.

## <span style="color:darkgreen;">Sorting the dataframe</span>

In [ ]:
pd??

In [14]:
# sort the data frame on basis of 'Age' values, by default the values will get sorted in ascending order
# Note: 'ascending = False' will sort the data frame in descending order.
# data1.sort_values('Age', ascending= True)
# inplace argument do it and apply to the same dataframe
data1.sort_values('Age', ascending= False, inplace=True)
data1

Age  Weight (in kg)  Height (in m)        BMI
8    75              23           1.24  14.958377
4    68              50           1.32  28.696051
20   68              50           1.32  28.696051
22   67              78           1.85  22.790358
18   67              78           1.85  22.790358
15   67              78           1.85  22.790358
7    57              34           1.61  13.116778
23   56              76           1.69  26.609713
14   56              76           1.69  26.609713
21   56              76           1.69  26.609713
17   56              74           1.69  25.909457
12   55              89           1.65  32.690542
2    54              78           1.50  34.666667
0    45              60           1.35  32.921811
11   34              65           1.76  20.983988
9    32              21           1.52   9.089335
16   26              65           1.21  44.395875
19   26              65           1.21  44.395875
3    26              65           1.21  44.395875
13   23              45           1.75  14.693878
10   23              53           1.50  23.555556
5    21              43           1.52  18.611496
1    12              43           1.21  29.369579
6    10              32           1.65  11.753903

In [8]:
round(data1['BMI'], 2)

0     32.92
1     29.37
2     34.67
3     44.40
4     28.70
5     18.61
6     11.75
7     13.12
8     14.96
9      9.09
10    23.56
11    20.98
12    32.69
13    14.69
14    26.61
15    22.79
16    44.40
17    25.91
18    22.79
19    44.40
20    28.70
21    26.61
22    22.79
23    26.61
Name: BMI, dtype: float64

In [9]:
data1['Age'] = data1['Age'].astype(int)

In [10]:
data1['Weight (in kg)'] = data1['Weight (in kg)'].astype(int)

In [13]:
data1.dtypes

Age                 int32
Weight (in kg)      int32
Height (in m)     float64
BMI               float64
dtype: object

In [ ]:
data1.assign(blank = None)

In [17]:
data1 =data1.reset_index(drop=True)

## <span style="color:darkgreen;">Droping Rows and Columns</span>

In [18]:
# create a copy of the dataframe
data3 = data1.copy()
data3

Age  Weight (in kg)  Height (in m)        BMI
0    75              23           1.24  14.958377
1    68              50           1.32  28.696051
2    68              50           1.32  28.696051
3    67              78           1.85  22.790358
4    67              78           1.85  22.790358
5    67              78           1.85  22.790358
6    57              34           1.61  13.116778
7    56              76           1.69  26.609713
8    56              76           1.69  26.609713
9    56              76           1.69  26.609713
10   56              74           1.69  25.909457
11   55              89           1.65  32.690542
12   54              78           1.50  34.666667
13   45              60           1.35  32.921811
14   34              65           1.76  20.983988
15   32              21           1.52   9.089335
16   26              65           1.21  44.395875
17   26              65           1.21  44.395875
18   26              65           1.21  44.395875
19   23              45           1.75  14.693878
20   23              53           1.50  23.555556
21   21              43           1.52  18.611496
22   12              43           1.21  29.369579
23   10              32           1.65  11.753903

In [33]:
# drop the BMI column
# axis 1 column, 0 is row, by default its 0 which is a column
# data3 =data1.drop('BMI', axis=1)
data3.drop('BMI', axis=1, inplace=True)

KeyError: "['BMI'] not found in axis"

In [20]:
data1

Age  Weight (in kg)  Height (in m)        BMI
0    75              23           1.24  14.958377
1    68              50           1.32  28.696051
2    68              50           1.32  28.696051
3    67              78           1.85  22.790358
4    67              78           1.85  22.790358
5    67              78           1.85  22.790358
6    57              34           1.61  13.116778
7    56              76           1.69  26.609713
8    56              76           1.69  26.609713
9    56              76           1.69  26.609713
10   56              74           1.69  25.909457
11   55              89           1.65  32.690542
12   54              78           1.50  34.666667
13   45              60           1.35  32.921811
14   34              65           1.76  20.983988
15   32              21           1.52   9.089335
16   26              65           1.21  44.395875
17   26              65           1.21  44.395875
18   26              65           1.21  44.395875
19   23              45           1.75  14.693878
20   23              53           1.50  23.555556
21   21              43           1.52  18.611496
22   12              43           1.21  29.369579
23   10              32           1.65  11.753903

In [31]:
data3

Age  Weight (in kg)  Height (in m)
0    75              23           1.24
1    68              50           1.32
2    68              50           1.32
3    67              78           1.85
4    67              78           1.85
5    67              78           1.85
6    57              34           1.61
7    56              76           1.69
8    56              76           1.69
9    56              76           1.69
10   56              74           1.69
11   55              89           1.65
12   54              78           1.50
13   45              60           1.35
14   34              65           1.76
15   32              21           1.52
16   26              65           1.21
17   26              65           1.21
18   26              65           1.21
19   23              45           1.75
20   23              53           1.50
21   21              43           1.52
22   12              43           1.21

In [28]:
# dropping the 23rd row
data3 = data1.drop(23)

Age  Weight (in kg)  Height (in m)        BMI
0    75              23           1.24  14.958377
1    68              50           1.32  28.696051
2    68              50           1.32  28.696051
3    67              78           1.85  22.790358
4    67              78           1.85  22.790358
5    67              78           1.85  22.790358
6    57              34           1.61  13.116778
7    56              76           1.69  26.609713
8    56              76           1.69  26.609713
9    56              76           1.69  26.609713
10   56              74           1.69  25.909457
11   55              89           1.65  32.690542
12   54              78           1.50  34.666667
13   45              60           1.35  32.921811
14   34              65           1.76  20.983988
15   32              21           1.52   9.089335
16   26              65           1.21  44.395875
17   26              65           1.21  44.395875
18   26              65           1.21  44.395875
19   23              45           1.75  14.693878
20   23              53           1.50  23.555556
21   21              43           1.52  18.611496
22   12              43           1.21  29.369579

In [32]:
data3.drop(23, inplace=True)

KeyError: '[23] not found in axis'

In [34]:
data3

Age  Weight (in kg)  Height (in m)
0    75              23           1.24
1    68              50           1.32
2    68              50           1.32
3    67              78           1.85
4    67              78           1.85
5    67              78           1.85
6    57              34           1.61
7    56              76           1.69
8    56              76           1.69
9    56              76           1.69
10   56              74           1.69
11   55              89           1.65
12   54              78           1.50
13   45              60           1.35
14   34              65           1.76
15   32              21           1.52
16   26              65           1.21
17   26              65           1.21
18   26              65           1.21
19   23              45           1.75
20   23              53           1.50
21   21              43           1.52
22   12              43           1.21

In [36]:
data3.shape

(23, 3)

## <span style="color:darkgreen;">Droping duplicates</span>

In [37]:
# Check if data has duplicates
data3.duplicated().sum()

7

In [38]:
# drop duplicates from your data
data3.drop_duplicates(inplace=True)


In [39]:
data3.duplicated().sum()

0

In [40]:
data3.shape

(16, 3)

## <span style="color:darkgreen;">Checking for missing values</span>

Let's import a new dataset.

In [41]:
# Import missingdata.csv 
mdata = pd.read_csv('missingdata.csv')
mdata.head(2)

Age  Weight (in kg)  Height (in m)
0  45.0            60.0           1.35
1  12.0            43.0           1.21

In [42]:
# check for nulls
mdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             22 non-null     float64
 1   Weight (in kg)  21 non-null     float64
 2   Height (in m)   22 non-null     float64
dtypes: float64(3)
memory usage: 680.0 bytes


In [43]:
mdata.isnull().sum()

Age               1
Weight (in kg)    2
Height (in m)     1
dtype: int64

The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.

## <span style="color:crimson;">Take home exercise</span>

<a id="reading_data"> </a>
### Reading Data from Different Sources

Note that the files names are used as examples only. You can try importing your own files to execute the below examples.

**1. Read a `.xlsx` file**

`pd.read_excel('example.xlsx')`

**2. Read a `.txt` file**

`data = pd.read_csv('example.txt', sep="\t")`

**3. Read a `.zip` file**

`import zipfile
with zipfile.ZipFile('data.zip') as z:
    with z.open('example.csv') as f:
        file = pd.read_csv(f)
        print(file.head())`

**4. Read a `.html` file**

`df = pd.read_html('example.html', header=1, index_col=0)`

**5. Read a `.json` file**

`pd.read_json('example.json')`